### IBM Applied Data Science Capstone Project

# Segmenting and Clustering Neighborhoods in London
In this notebook, neighborhoods in the city of London are explored, segmented, and clustered. For the London neighborhood data, a Wikipedia page exists that has all the information needed to explore and cluster the neighborhoods in London. The data is scraped from the Wikipedia page and wrangled, cleaned and then read into a pandas dataframe so that it is in a structured format.

Once the data is in a structured format, Analyze to open a Japanese restaurant and where would we recommend that to open it? 

# Methodology

Install the required packages.

In [3]:
!pip install arcgis
!pip install wikipedia
!conda install -c conda-forge geopy --yes
!pip install geocoder
!pip install folium
print('Libraries Installed.')

Libraries Installed.


### Importing the required packages.

In [4]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import geocoder # to get coordinates
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

import wikipedia as wp

import folium # map rendering library

from arcgis.geocoding import geocode
from arcgis.gis import GIS
gis = GIS()
print('Libraries imported.')

Libraries imported.


The task is to explore the city and plot the map to show the Neighborhoods being considered and then build our model by clustering all of the similar Neighborhoods together and finally plot the new map with the clustered Neighborhoods.

## 1. Web-scrape and Explore Dataset
### Exploring London City

### Neighborhoods of London

Collecting data needed for the our business solution from Wiki.

### Data Collection

In [12]:
#Get the html source
html = wp.page("List of areas of London").html().encode("UTF-8")
df = pd.read_html(html, flavor='html5lib')[1]     
df.head()

Location                     London borough       Post town  \
0   Abbey Wood              Bexley, Greenwich [7]          LONDON   
1        Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2    Addington                         Croydon[8]         CROYDON   
3   Addiscombe                         Croydon[8]         CROYDON   
4  Albany Park                             Bexley  BEXLEY, SIDCUP   

  Postcode district Dial code OS grid ref  
0               SE2       020    TQ465785  
1            W3, W4       020    TQ205805  
2               CR0       020    TQ375645  
3               CR0       020    TQ345665  
4         DA5, DA14       020    TQ478728

### Data Preprocessing

In [6]:
df.rename(columns=lambda x: x.strip().replace(" ", "_"), inplace=True)
df.head()

Location                     London borough       Post_town  \
0   Abbey Wood              Bexley, Greenwich [7]          LONDON   
1        Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2    Addington                         Croydon[8]         CROYDON   
3   Addiscombe                         Croydon[8]         CROYDON   
4  Albany Park                             Bexley  BEXLEY, SIDCUP   

  Postcode district Dial code OS_grid_ref  
0               SE2       020    TQ465785  
1            W3, W4       020    TQ205805  
2               CR0       020    TQ375645  
3               CR0       020    TQ345665  
4         DA5, DA14       020    TQ478728

### Feature Selection
Keep only relavant boroughs, Post town and district for further steps.

In [7]:
df1 = df.drop( [ df.columns[0], df.columns[4], df.columns[5] ], axis=1)
df1.columns = ['borough','town','post_code']
df1['borough'] = df1['borough'].map(lambda x: x.rstrip(']').rstrip('0123456789').rstrip('['))
df1.head()

borough            town  post_code
0              Bexley, Greenwich           LONDON        SE2
1  Ealing, Hammersmith and Fulham          LONDON     W3, W4
2                         Croydon         CROYDON        CR0
3                         Croydon         CROYDON        CR0
4                          Bexley  BEXLEY, SIDCUP  DA5, DA14

Dimension of the dataframe

In [8]:
df1.shape

(531, 3)

We currently have 531 records and 3 columns of our data. Lets do the Feature Engineering

In [9]:
df1 = df1[df1['town'].str.contains('LONDON')]
df1

borough  \
0                                Bexley, Greenwich    
1                    Ealing, Hammersmith and Fulham   
6                                              City   
7                                       Westminster   
9                                           Bromley   
10                                        Islington   
12                                        Islington   
14                                           Barnet   
15                                          Enfield   
16                                       Wandsworth   
17                                        Southwark   
18                                             City   
22                             Richmond upon Thames   
24                                           Barnet   
26                                        Islington   
27                                       Wandsworth   
28                                      Westminster   
29                                          Bromley   
30                                           Newham   
34                                           Ealing   
35                                      Westminster   
36                                         Lewisham   
39                                           Camden   
41                                        Southwark   
43                                    Tower Hamlets   
45                                           Bexley   
49                                             City   
50                                         Lewisham   
51                                        Greenwich   
52                                    Tower Hamlets   
54                                           Camden   
56                                         Haringey   
57                                    Tower Hamlets   
58                                         Haringey   
60                                           Barnet   
61                                            Brent   
63                                          Lambeth   
64                                         Lewisham   
66                                    Tower Hamlets   
68   Kensington and Chelsea, Hammersmith and Fulham   
69                                            Brent   
70                                           Barnet   
73                                           Barnet   
74                                        Southwark   
75                                    Tower Hamlets   
76                                           Camden   
77                                    Tower Hamlets   
78                                   Waltham Forest   
79                                           Newham   
80                                        Islington   
82                             Richmond upon Thames   
84                                         Lewisham   
86                                           Camden   
87                                      Westminster   
88                                        Greenwich   
91                           Kensington and Chelsea   
94                                           Barnet   
95                                      Westminster   
96                                         Lewisham   
97                                   Waltham Forest   
99         Hounslow, Ealing, Hammersmith and Fulham   
100                                           Brent   
101                                          Barnet   
102                             Lambeth, Wandsworth   
103                                       Islington   
105                                          Barnet   
107                                          Merton   
108                                          Barnet   
114                                     Westminster   
121                           Barnet, Brent, Camden   
122                                        Lewisham   
124                                          Bexley   
125                          

In [10]:
df1.shape

(308, 3)

We now have only 308 rows. We can proceed with our further steps. Getting some descriptive statistics

# References:
* [London Areas Wiki](https://en.wikipedia.org/wiki/List_of_areas_of_London)
* [Foursquare API](https://foursquare.com/)
* [ArcGIS API](https://www.arcgis.com/index.html)

<hr>

### Thank You